workspaces/optic-engine/src/queries/shape.rs

```diff
@@ -438,6 +441,96 @@ impl<'a> ShapeQueries<'a> {
     let command = ShapeCommand::remove_field(field_node.field_id.clone());
     Some(std::iter::once(command))
   }
+
+  pub fn edit_shape_commands(
+    &'a self,
+    shape_trail: &ShapeTrail,
+    required_kinds: impl IntoIterator<Item = &'a ShapeKind>,
+    id_generator: &mut impl SpecIdGenerator,
+  ) -> Option<impl Iterator<Item = ShapeCommand>> {
```

Comment:
The magic happens here. For now, we're straight up accepting `ShapeKind` instances to detail what a shape should be. For nullable and optionals, that's enough. For more complex types, it might not be, but since those are out of the scope of the first version of field level editing, lets cross that bridge when we get to it.

Reply:
That might well very be how that's going to shape out, I'm not 100% sure yet. Regardless, this interface is something we can further iterate on, or leave as is. It does the most core thing without any frills at the moment, so should be a good base either way.